agent/reconcile.go

```diff
@@ -275,51 +270,45 @@ func (ar *AgentReconciler) calculateTaskChainForUnit(dState *AgentState, cState
 		return nil
 	}
 
-	tc := newTaskChain(u)
 	if *cJState == job.JobStateInactive {
-		tc.Add(task{
+		tasks = append(tasks, task{
 			typ:    taskTypeLoadUnit,
 			reason: taskReasonScheduledButUnloaded,
+			unit:   u,
 		})
 	}
 
 	if (*cJState == job.JobStateInactive || *cJState == job.JobStateLoaded) && dJob.TargetState == job.JobStateLaunched {
-		tc.Add(task{
+		tasks = append(tasks, task{
 			typ:    taskTypeStartUnit,
 			reason: taskReasonLoadedDesiredStateLaunched,
+			unit:   u,
 		})
 	}
 
 	if *cJState == job.JobStateLaunched && dJob.TargetState == job.JobStateLoaded {
-		tc.Add(task{
+		tasks = append(tasks, task{
 			typ:    taskTypeStopUnit,
 			reason: taskReasonLaunchedDesiredStateLoaded,
+			unit:   u,
 		})
 	}
 
-	if len(tc.tasks) == 0 {
+	if len(tasks) == 0 {
 		log.Errorf("Unable to determine how to reconcile Job(%s): desiredState=%#v currentState=%#v", jName, dJob, cJState)
-		return nil
 	}
 
-	return &tc
+	return
 }
 
-func (ar *AgentReconciler) launchTaskChain(tc taskChain, a *Agent) {
-	log.Debugf("AgentReconciler attempting task chain %s", tc)
-	reschan, err := ar.tManager.Do(tc, a)
-	if err != nil {
-		log.Infof("AgentReconciler task chain failed: chain=%s err=%v", tc, err)
-		return
-	}
-
-	go func() {
```

Comment:
It's sort of a long story. The short version is that it used to be safe to spin off a goroutine here because the TaskManager would reject new taskChains for units that already have work in progress. Now that the TaskManager does not track the in-flight tasks (and therefore is no longer threadsafe), the reconciler should not interact with the TaskManager concurrently (which this goroutine would allow).